cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -1662,7 +1662,348 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
+                                RegMem::reg(lhs),
+                                tmp_reg,
+                            ));
+
+                            ctx.emit(Inst::gen_move(dst, rhs, output_ty));
+                            ctx.emit(Inst::xmm_rm_r_imm(
+                                SseOpcode::Palignr,
+                                RegMem::reg(rhs),
+                                dst,
+                                8,
+                                OperandSize::Size32,
+                            ));
+                            ctx.emit(Inst::xmm_mov(SseOpcode::Pmovsxbw, RegMem::reg(rhs), dst));
+                            ctx.emit(Inst::xmm_rm_r(
+                                SseOpcode::Pmullw,
+                                RegMem::reg(tmp_reg.to_reg()),
+                                dst,
+                            ));
+                        }
+                        (types::I16X8, types::I16X8, types::I32X4) => {
+                            // i32x4.extmul_high_i16x8_s
+                            ctx.emit(Inst::gen_move(dst, lhs, input0_ty));
+                            let tmp_reg = ctx.alloc_tmp(types::I16X8).only_reg().unwrap();
+                            ctx.emit(Inst::gen_move(tmp_reg, lhs, input0_ty));
+                            ctx.emit(Inst::xmm_rm_r(SseOpcode::Pmullw, RegMem::reg(rhs), dst));
+                            ctx.emit(Inst::xmm_rm_r(SseOpcode::Pmulhw, RegMem::reg(rhs), tmp_reg));
+                            ctx.emit(Inst::xmm_rm_r(
+                                SseOpcode::Punpckhwd,
+                                RegMem::from(tmp_reg),
+                                dst,
+                            ));
+                        }
+                        (types::I32X4, types::I32X4, types::I64X2) => {
+                            // i64x2.extmul_high_i32x4_s
+                            let tmp_reg = ctx.alloc_tmp(types::I32X4).only_reg().unwrap();
+                            ctx.emit(Inst::xmm_rm_r_imm(
+                                SseOpcode::Pshufd,
+                                RegMem::reg(lhs),
+                                tmp_reg,
+                                0xFA,
+                                OperandSize::Size32,
+                            ));
+                            ctx.emit(Inst::xmm_rm_r_imm(
+                                SseOpcode::Pshufd,
+                                RegMem::reg(rhs),
+                                dst,
+                                0xFA,
+                                OperandSize::Size32,
+                            ));
+                            ctx.emit(Inst::xmm_rm_r(
+                                SseOpcode::Pmuldq,
+                                RegMem::reg(tmp_reg.to_reg()),
+                                dst,
+                            ));
+                        }
+                        _ => panic!("Unsupported extmul_low_signed type"),
```

Comment:
Initially I thought that this matching was too greedy--that it would match sequences that it shouldn't be and then crash because the types are wrong. I think now that it is OK because `swiden_high` only really allows the following types: `I8X16`, `I16X8`, `I32X4`. I think it might be worthwhile to mention this assumption in some comments at the top so that future readers aren't confused.

```
v0 = swiden.i64
```